### PR TITLE
add various channel mode tests

### PR DESCRIPTION
**Pull request by slingamn**
Blocked for Ergo by https://github.com/ergochat/ergo/pull/2163 but I wanted to see what the state of play is with other ircds

### CODE REVIEW

**Comment by slingamn** [comment]
* Sable and ircu don't send 329 RPL_CREATIONTIME
* Inspircd sends no response for `MODE #channel +o existing_but_not_present_nick` (it sends 401 ERR_NOSUCHNICK if the nick doesn't exist); this is arguably a bug
* Unreal sends 401 ERR_NOSUCHNICK for `MODE #nonexistentchannel +o existingnick`; Modern specifies 403 ERR_NOSUCHCHANNEL here but Unreal might be following some older spec

**Comment by slingamn** [comment]
Unreal also sends an extra parameter for 324 RPL_CHANNELMODEIS:

    :My.Little.Server 324 chanop #chan +intH 100:1d

I think that's the argument to `+H` (history)? Not sure about the specification basis for this.

File: irctest/server_tests/chmodes/modeis.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+from irctest.numerics import RPL_CHANNELCREATED, RPL_CHANNELMODEIS
+from irctest.patma import ANYSTR, ListRemainder
+
+
+class RplChannelModeIsTestCase(cases.BaseServerTestCase):
+    @cases.mark_specifications("Modern")
+    @cases.xfailIfSoftware(
+        ["irc2", "Sable"],
+        "irc2 doesn't support 329 RPL_CHANNELCREATED"
+        "https://github.com/Libera-Chat/sable/issues/130",
```
Comment (progval) [inline irctest/server_tests/chmodes/modeis.py:18]:
```suggestion
        "irc2 doesn't support 329 RPL_CHANNELCREATED "
        "https://github.com/Libera-Chat/sable/issues/130",
```

**Comment by slingamn** [comment]
The Inspircd fix should be in their 4.0 release, which should be coming soon-ish

**Comment by slingamn** [comment]
@progval I think this is finally done! I updated the Insp stable pointer to `v4.0.0rc1` to pick up the fix, so we can either merge this, or hold it until the official 4.0 release and merge it then.

**Comment by slingamn** [comment]
Hmm, Atheme 7.2.12 is incompatible with Insp 4:

```
2024-06-10T01:27:16.9371515Z ----------------------------- Captured stderr call -----------------------------
2024-06-10T01:27:16.9372270Z [2024-06-10 01:18:38] atheme 7.2.12 is starting up...
2024-06-10T01:27:16.9372873Z [2024-06-10 01:18:38] module_locate_symbol(): nickserv/set_core is not loaded.
2024-06-10T01:27:16.9373788Z [2024-06-10 01:18:38] db-open-read: database '/tmp/tmpzfsfo97a/services.db' does not yet exist; a new one will be created.
2024-06-10T01:27:16.9374503Z [2024-06-10 01:18:38] pid 2071
2024-06-10T01:27:16.9374906Z [2024-06-10 01:18:38] running in foreground mode from /usr
2024-06-10T01:27:16.9375595Z [2024-06-10 01:18:38] uplink_connect(): connecting to first entry My.Little.Server[0.0.0.0]:41335.
2024-06-10T01:27:16.9376338Z [2024-06-10 01:18:38] irc_handle_connect(): connection to uplink established
2024-06-10T01:27:16.9377786Z [2024-06-10 01:18:38] m_error(): error from server: CAPAB negotiation failed: Server is using protocol version 1202 which is too old to link with this server (protocol versions 1205 to 1206 are supported)
2024-06-10T01:27:16.9379080Z [2024-06-10 01:18:48] uplink_connect(): trying again first entry My.Little.Server[0.0.0.0]:41335
2024-06-10T01:27:16.9379830Z [2024-06-10 01:18:48] irc_handle_connect(): connection to uplink established
2024-06-10T01:27:16.9380451Z --------------------------- Captured stderr teardown ---------------------------
2024-06-10T01:27:16.9381674Z [2024-06-10 01:18:48] m_error(): error from server: CAPAB negotiation failed: Server is using protocol version 1202 which is too old to link with this server (protocol versions 1205 to 1206 are supported)
2024-06-10T01:27:16.9382833Z [2024-06-10 01:18:48] wallops(): unable to send: Exiting on signal 15.
2024-06-10T01:27:16.9383400Z [2024-06-10 01:18:48] sighandler(): got SIGTERM; exiting...
2024-06-10T01:27:16.9383850Z [2024-06-10 01:18:48] main(): shutting down
```

**Comment by slingamn** [comment]
Looks like this is blocked: https://github.com/atheme/atheme/issues/904

**Comment by progval** [comment]
I'll write a PR to add `inspircd-3` as its own software (like `unrealircd-5`) and remove atheme from insp 4

**Comment by slingamn** [comment]
Looks like we hit a real, but nondeterministic, bug in Sable:

```
2024-06-10T05:43:08.9214996Z ----------------------------- Captured stderr call -----------------------------
2024-06-10T05:43:08.9215840Z thread 'tokio-runtime-worker' panicked at sable_ircd/src/server/mod.rs:501:29:
2024-06-10T05:43:08.9216437Z Lost auth client task
2024-06-10T05:43:08.9217002Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2024-06-10T05:43:08.9217786Z Error: Custom { kind: UnexpectedEof, error: "could not read" }
```

Should we turn on backtraces?

**Comment by progval** [comment]
I'll debug it locally

**Comment by progval** [comment]
Oh that was just one of Sable's regular flaky failures. I still don't know how to fix those besides restarting the test.

File: irctest/server_tests/chmodes/operator.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        messages = self.getMessages("chanop")
+        # ERR_NOSUCHNICK is typical, Bahamut additionally sends ERR_USERNOTINCHANNEL
+        if self.controller.software_name != "Bahamut":
+            self.assertEqual(len(messages), 1)
+            self.assertMessageMatch(messages[0], command=ERR_NOSUCHNICK)
+        else:
+            self.assertLessEqual(len(messages), 2)
+            commands = {message.command for message in messages}
+            self.assertLessEqual({ERR_NOSUCHNICK}, commands)
+            self.assertLessEqual(commands, {ERR_NOSUCHNICK, ERR_USERNOTINCHANNEL})
```
Comment (progval) [inline irctest/server_tests/chmodes/operator.py:72]: https://github.com/DALnet/bahamut/pull/241

File: irctest/server_tests/chmodes/operator.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        self.sendLine("chanop", "MODE #nonexistentchan +o chanop")
+        messages = self.getMessages("chanop")
+        self.assertEqual(len(messages), 1)
+        # Modern: "If <target> is a channel that does not exist on the network,
+        # the ERR_NOSUCHCHANNEL (403) numeric is returned."
+        # However, Unreal and ngircd send 401 ERR_NOSUCHNICK here instead:
+        if self.controller.software_name not in ("UnrealIRCd", "ngIRCd"):
+            self.assertEqual(messages[0].command, ERR_NOSUCHCHANNEL)
+        else:
+            self.assertIn(messages[0].command, [ERR_NOSUCHCHANNEL, ERR_NOSUCHNICK])
```
Comment (progval) [inline irctest/server_tests/chmodes/operator.py:88]: https://github.com/ngircd/ngircd/pull/319
https://github.com/unrealircd/unrealircd/pull/287

**Comment by SadieCat** [comment]
> The Inspircd fix should be in their 4.0 release, which should be coming soon-ish

This will also be in our next v3 release which comes out the Friday after next.

**Pull request merged**